### ISSUE TITLE
systemd WorkingDirectory paths mounted too late

### ISSUE THREAD

**Issue by lnicola** [dasJ/sd-zfs#23]
I have a `systemd` service with a custom `WorkigDirectory` pointing to a `zfs` filesystem from another pool (i.e. not the root one). It fails on boot with `EXIT_CHDIR` because the path isn't accessible yet. I would expect the service to start:

> Units with WorkingDirectory=, RootDirectory=, RootImage=, RuntimeDirectory=, StateDirectory=, CacheDirectory=, LogsDirectory= or ConfigurationDirectory= set automatically gain dependencies of type Requires= and After= on all mount units required to access the specified paths. This is equivalent to having them listed explicitly in RequiresMountsFor=.

However, things are much better now than in `1.0`. I have other services that either require or are located filesystems on the root pool. Previously, they didn't start, but `systemd` reported no error and showed them as "loaded, not running".

**Comment by dasJ** [comment]
Well - I never expected anyone to use this in their initrd. Can you add `systemd-analyze` to your initrd and run `systemd-analyze dump` from there? I only need the details of the unit you wrote and any mount units it depends on. Also, a `systemctl cat` (don't forget to add `systemctl` to the initrd) of the mount units (if any) would be extremely helpful.

**Comment by lnicola** [comment]
I'm not really familiar with the boot process, but I want to add that my service isn't present in the `initramfs`. It's in the "real" system that gets booted.

Am I missing something?

**Comment by lnicola** [comment]
`systemctl show` says that my service has `After` and `Requires` directives on the `fs.mount` unit it needs. It also has `RequiresMountsFor` set to that path. I'm not sure why it doesn't work, then.

**Comment by dasJ** [comment]
Yep, sd-zfs is only run in the initrd. When your system switches to the real root with your real data, sd-zfs is done and will not be used again. Normally, the initrd mounts prepares everything for the real init to run (decrypt disks, mount partitions like / and /usr). So, this needs to be fixed in your normal systemd system instance. Can you give me the `systemctl cat` if your `fs.mount`? Also, did you write that unit or was it generated?

**Comment by lnicola** [comment]
It's a unit I've written. `fs.mount` doesn't work in `systemctl cat`.

```
$ systemctl show tank-storage.mount
Where=/tank/storage
What=tank/storage
Options=rw,noatime,xattr,posixacl
Type=zfs
TimeoutUSec=1min 30s
ControlPID=0
DirectoryMode=0755
SloppyOptions=no
LazyUnmount=no
ForceUnmount=no
Result=success
UID=[not set]
GID=[not set]
Slice=system.slice
ControlGroup=/system.slice/tank-storage.mount
MemoryCurrent=[not set]
CPUUsageNSec=[not set]
TasksCurrent=0
IPIngressBytes=18446744073709551615
IPIngressPackets=18446744073709551615
IPEgressBytes=18446744073709551615
IPEgressPackets=18446744073709551615
Delegate=no
CPUAccounting=no
CPUWeight=[not set]
StartupCPUWeight=[not set]
CPUShares=[not set]
StartupCPUShares=[not set]
CPUQuotaPerSecUSec=infinity
IOAccounting=no
IOWeight=[not set]
StartupIOWeight=[not set]
BlockIOAccounting=no
BlockIOWeight=[not set]
StartupBlockIOWeight=[not set]
MemoryAccounting=no
MemoryLow=0
MemoryHigh=infinity
MemoryMax=infinity
MemorySwapMax=infinity
MemoryLimit=infinity
DevicePolicy=auto
TasksAccounting=yes
TasksMax=4915
IPAccounting=no
UMask=0022
LimitCPU=infinity
LimitCPUSoft=infinity
LimitFSIZE=infinity
LimitFSIZESoft=infinity
LimitDATA=infinity
LimitDATASoft=infinity
LimitSTACK=infinity
LimitSTACKSoft=8388608
LimitCORE=infinity
LimitCORESoft=infinity
LimitRSS=infinity
LimitRSSSoft=infinity
LimitNOFILE=4096
LimitNOFILESoft=1024
LimitAS=infinity
LimitASSoft=infinity
LimitNPROC=61719
LimitNPROCSoft=61719
LimitMEMLOCK=16777216
LimitMEMLOCKSoft=16777216
LimitLOCKS=infinity
LimitLOCKSSoft=infinity
LimitSIGPENDING=61719
LimitSIGPENDINGSoft=61719
LimitMSGQUEUE=819200
LimitMSGQUEUESoft=819200
LimitNICE=0
LimitNICESoft=0
LimitRTPRIO=0
LimitRTPRIOSoft=0
LimitRTTIME=infinity
LimitRTTIMESoft=infinity
OOMScoreAdjust=0
Nice=0
IOSchedulingClass=0
IOSchedulingPriority=0
CPUSchedulingPolicy=0
CPUSchedulingPriority=0
TimerSlackNSec=50000
CPUSchedulingResetOnFork=no
NonBlocking=no
StandardInput=null
StandardOutput=inherit
StandardError=inherit
TTYReset=no
TTYVHangup=no
TTYVTDisallocate=no
SyslogPriority=30
SyslogLevelPrefix=yes
SyslogLevel=6
SyslogFacility=3
SecureBits=0
CapabilityBoundingSet=cap_chown cap_dac_override cap_dac_read_search cap_fowner cap_fsetid cap_kill cap_setgid cap_setuid cap_setpcap cap_linux_immutable cap_net_bind_service cap_net_broadcast cap_net_admin cap_net_raw cap_ipc_lock cap_ipc_owner cap_sys_module cap_sys_rawio cap_sys_chroot cap_sys_ptrace cap_sys_pacct cap_sys_admin cap_sys_boot cap_sys_nice cap_sys_resource cap_sys_time cap_sys_tty_config cap_mknod cap_lease cap_audit_write cap_audit_control cap_setfcap cap_mac_override cap_mac_admin cap_syslog cap_wake_alarm cap_block_suspend
AmbientCapabilities=
DynamicUser=no
RemoveIPC=no
MountFlags=
PrivateTmp=no
PrivateDevices=no
ProtectKernelTunables=no
ProtectKernelModules=no
ProtectControlGroups=no
PrivateNetwork=no
PrivateUsers=no
ProtectHome=no
ProtectSystem=no
SameProcessGroup=yes
UtmpMode=init
IgnoreSIGPIPE=yes
NoNewPrivileges=no
SystemCallErrorNumber=0
LockPersonality=no
RuntimeDirectoryPreserve=no
RuntimeDirectoryMode=0755
StateDirectoryMode=0755
CacheDirectoryMode=0755
LogsDirectoryMode=0755
ConfigurationDirectoryMode=0755
MemoryDenyWriteExecute=no
RestrictRealtime=no
RestrictNamespaces=no
MountAPIVFS=no
KeyringMode=private
KillMode=control-group
KillSignal=15
SendSIGKILL=yes
SendSIGHUP=no
Id=tank-storage.mount
Names=tank-storage.mount
Requires=system.slice
RequiredBy=rusty-share.service
Conflicts=umount.target
Before=umount.target rusty-share.service local-fs.target
After=-.mount local-fs-pre.target system.slice
RequiresMountsFor=/tank
Description=/tank/storage
LoadState=loaded
ActiveState=active
SubState=mounted
SourcePath=/proc/self/mountinfo
StateChangeTimestamp=Fri 2017-11-10 19:44:45 EET
StateChangeTimestampMonotonic=16588984
InactiveExitTimestamp=Fri 2017-11-10 19:44:45 EET
InactiveExitTimestampMonotonic=16588984
ActiveEnterTimestamp=Fri 2017-11-10 19:44:45 EET
ActiveEnterTimestampMonotonic=16588984
ActiveExitTimestampMonotonic=0
InactiveEnterTimestampMonotonic=0
CanStart=yes
CanStop=yes
CanReload=yes
CanIsolate=no
StopWhenUnneeded=no
RefuseManualStart=no
RefuseManualStop=no
AllowIsolate=no
DefaultDependencies=yes
OnFailureJobMode=replace
IgnoreOnIsolate=yes
NeedDaemonReload=no
JobTimeoutUSec=infinity
JobRunningTimeoutUSec=infinity
JobTimeoutAction=none
ConditionResult=no
AssertResult=no
ConditionTimestampMonotonic=0
AssertTimestampMonotonic=0
Transient=no
Perpetual=no
StartLimitIntervalSec=10000000
StartLimitBurst=5
StartLimitAction=none
InvocationID=e3c02fb15b674b2a9932a8b22addfb1e
```

```
$ systemctl cat rusty-share              
# /etc/systemd/system/rusty-share.service
[Unit]
Description=Rusty Share
After=network.target
After=tank-storage.mount
Requires=tank-storage.mount

[Service]
User=http
ExecStart=/usr/local/bin/rusty-share
WorkingDirectory=/tank/storage

[Install]
WantedBy=default.target
```

**Comment by dasJ** [comment]
Strange - systemd should execute something like `mount -t zfs -o rw,noatime,xattr,posixacl tank/storage /tank/storage`. Does this command (and exactly this command) work for you after unmounting the storage?

**Comment by dasJ** [comment]
Also, try adding `After=zfs-import-cache.service zfs-import-scan.service` to the mount unit

**Comment by lnicola** [comment]
```shell
$ sudo mount -t zfs -o rw,noatime,xattr,posixacl tank/storage /tank/storage
filesystem 'tank/storage' cannot be mounted using 'mount'.
Use 'zfs set mountpoint=legacy' or 'zfs mount tank/storage'.
See zfs(8) for more information.
```

~~I noticed that I still have `zfs-mount.service` enabled. It's failing now, because `/tank/storage` is already mounted, but I suspect it might paper over the problem if `sd-zfs` doesn't work. I'd rather leave it enabled because I also keep the standard `initramfs` image just in case.~~

Also, possibly related to the latest `zfs-utils-common` updates, my service is now starting on boot. It's either been fixed, or there's a race condition somewhere.

EDIT: I had booted with the other `initramfs` by mistake. However, things are still working as expected.
